packages/datadog-instrumentations/src/kafkajs.js

```diff
@@ -103,39 +107,72 @@ addHook({ name: 'kafkajs', file: 'src/index.js', versions: ['>=1.4'] }, (BaseKaf
     const run = consumer.run
 
     const groupId = arguments[0].groupId
-    consumer.run = function ({ eachMessage, ...runArgs }) {
-      if (typeof eachMessage !== 'function') return run({ eachMessage, ...runArgs })
-
+    consumer.run = function ({ eachMessage, eachBatch, ...runArgs }) {
       return run({
-        eachMessage: function (...eachMessageArgs) {
-          const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
-          return innerAsyncResource.runInAsyncScope(() => {
-            const { topic, partition, message } = eachMessageArgs[0]
-            consumerStartCh.publish({ topic, partition, message, groupId })
-            try {
-              const result = eachMessage.apply(this, eachMessageArgs)
-              if (result && typeof result.then === 'function') {
-                result.then(
-                  innerAsyncResource.bind(() => consumerFinishCh.publish(undefined)),
-                  innerAsyncResource.bind(err => {
-                    if (err) {
-                      consumerErrorCh.publish(err)
-                    }
-                    consumerFinishCh.publish(undefined)
-                  })
-                )
-              } else {
+        eachMessage: typeof eachMessage === 'function'
+          ? function (...eachMessageArgs) {
+            const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
+            return innerAsyncResource.runInAsyncScope(() => {
+              const { topic, partition, message } = eachMessageArgs[0]
+              consumerStartCh.publish({ topic, partition, message, groupId })
+              try {
+                const result = eachMessage.apply(this, eachMessageArgs)
+                if (result && typeof result.then === 'function') {
+                  result.then(
+                    innerAsyncResource.bind(() => consumerFinishCh.publish(undefined)),
+                    innerAsyncResource.bind(err => {
+                      if (err) {
+                        consumerErrorCh.publish(err)
+                      }
+                      consumerFinishCh.publish(undefined)
+                    })
+                  )
+                } else {
+                  consumerFinishCh.publish(undefined)
+                }
+
+                return result
+              } catch (e) {
+                consumerErrorCh.publish(e)
                 consumerFinishCh.publish(undefined)
+                throw e
               }
-
-              return result
-            } catch (e) {
-              consumerErrorCh.publish(e)
-              consumerFinishCh.publish(undefined)
-              throw e
+            })
+          }
+          : eachMessage,
+        eachBatch:
+          typeof eachBatch === 'function'
+            ? function (...eachBatchArgs) {
+              const innerAsyncResource = new AsyncResource('bound-anonymous-fn')
+              return innerAsyncResource.runInAsyncScope(() => {
```

Comment:
This is not how eachBatch works. The eachBatch callback provided to `consumer.run` will run once all messages are consumed, and it will provide the consumed messages to the callback defined with eachBatch. 

With that said, we have discussed this internally among IDM team, and we think it would be most appropriate to create 1 span for the entire batch. Additionally, the context propagation of parent-child was specifically requested by a customer, and for that reason, we are adding that propagation config hidden behind a disabled feature flag by default. We will document that this propagation is 'experimental', and that only the first valid context found within a message will be used as the parent context. 

Ideally, we would like to use span links (but they are not ready and won't be for a little while), but this feature has been continually requested by a customer and we need it implemented.

Reply:
> this feature has been continually requested by a customer and we need it implemented

Yes, it has been continually requested to be implemented but properly, where each message gets its own continuity, including downstream. We do support only the first span in other integrations, and it has only led to confusion as it not only provides random bits of information, but also it results in incorrect parent-child relationship. Reverting this incorrect behaviour would then be considered a breaking change, so it shouldn't be added lightly.